Feature/min max improvements

micw opened the pull request:
## Description:


**Related issue (if applicable):** fixes 	Fix #6783
**Related issue (if applicable):** fixes 	Fix #6784

## Checklist:

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been modified to verify that the new code works.

micw committed:
Fix #6783, remove a test that makes no sense anymore

micw committed:
Fix #6784

pavoni commented:
@stu-gott Thanks for your super fast feedback - can you double check this still works with your CoffeeMachine?

exxamalte commented:
The changes look all fine to me. Good improvements.

fabaff commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -150,10 +152,38 @@ def async_update(self):
+    # pylint: disable=R0201
+    def calc_min(self, sensor_values):
+        """Calculate min value, honoring unkown states."""
+        val = STATE_UNKNOWN
```
This cannot happen as we are filtering `STATE_UNKNOWN` in the listener.

micw replied:
Right, I'll add a test case and fix it. Since STATE_UNKNOWN is now handled, it should not be filtered in the listener any more.

micw replied:
fixed

fabaff commented:
This would also require a documentation update.

micw commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -76,9 +74,13 @@ def __init__(self, hass, entity_ids, name, sensor_type, round_digits):
+        else:
+            self._name = '{} sensor'.format(
+                next(v for k, v in SENSOR_TYPES.items()
+                     if self._sensor_type == v)).capitalize()
```
How to deal with such?

micw replied:
(answer from gitter): Just need to mention it in release notes

micw commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -98,7 +100,7 @@ def async_min_max_sensor_state_listener(entity, old_state, new_state):
             if self._unit_of_measurement != new_state.attributes.get(
                     ATTR_UNIT_OF_MEASUREMENT):
                 _LOGGER.warning("Units of measurement do not match")
-                return
```
Agree but stopping the update is also a good solution. Let's discuss this first on gitter?

micw replied:
Update: when reloading configuration, unit_of_measurement may change - either the one configured with this sensor or those configured with the calculated entities. So the following situation can happen with the old code (correct me, if i'm wrong)
- first incoming unit is wrong and remembered by this sensor. when it's fixed in config and config is reloaded, it won't work

Suggestions to solve this:
- ignore units, use only the unig configured for min_max
- remember units for all values. if all matches, use this unit. otherwise return ERROR as unit, that's clearly visible in the UI

fabaff replied:
> either the one configured with this sensor

The `min_max` sensor has no option to set the `unit_of_measurement` by itself.

> first incoming unit is wrong and remembered by this sensor. when it's fixed in config and config is reloaded, it won't work

If you get a unit of measurement mismatch and fix that then the sensor will work. The `min_max` sensor like all other sensors don't know what was stored before the HA restart.

> ignore units, use only the unig configured for min_max

As I told before, not a smart way to go as the result will become unpredictable. If you want to do that you can use a `template` sensor first and feed that to the `min_max` sensor.

> remember units for all values. if all matches, use this unit. otherwise return ERROR as unit, that's clearly visible in the UI

I decided to use the first entity as lead for the unit of measurement. Sure if the first one is not what you want the sensor assume that the rest of the entities is wrong. Humans can deal with false-positives pretty quick.

Well, the log file is for errors and warning. We could create a persistent notification and add the `entity_id` to the warning. Add the `entity_id` would be an improvement anyway.

micw replied:
I had a look into persistent notification and came to the conclusion that it's a bit overkill.
I just re-implemented handling of mismatch in units and set value to unknown plus unit to ERR. I have re-added your test case for this,

micw commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -150,10 +152,38 @@ def async_update(self):
+        self.max_value = self.calc_max(sensor_values)
+        self.mean = self.calc_mean(sensor_values)
+
+    # pylint: disable=R0201
```
Any hint, how to? I tried (including google search) but got no solution that worked with lint.

micw replied:
update: fixed.

fabaff replied:
Looks like that the `@staticmethod` decorator was missing.

micw commented:
If you basically agree with my changes (althought it's breaking), I will commit a PR to the docs as well.
Since we already have a breaking change, I would also switch the property "entitiy_ids" to "entities" to align it with "groups" component.

micw committed:
Fix typo in docstring

micw committed:
Fix handling of known->unknown state, extended test, fix lint errors

micw commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -150,10 +152,38 @@ def async_update(self):
+                    val = sval
+        return val
+
+    # pylint: disable=R0201
```
fixed

micw commented:
@fabaff Could your please proceed with this PR?

micw committed:
Better handling of mismatch in unit of measurement.

Set state to "unkown" and unit of measurement to "ERR" if unit of measurement differs between aggregatet states.
Add entity_id to logged warning.

micw committed:
Make icon configurable

micw committed:
Fix typo

micw committed:
Fix lint

micw committed:
Fix lint

micw commented:
@fabaff Do you think we can go with the changes I made? I's leave entity_ids as it is.

micw committed:
Add option to set entity_id on min_max sensor

micw committed:
Fix lint

micw committed:
Fix lint logging-not-lazy

fabaff commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -47,9 +47,11 @@
+    vol.Optional(CONF_ID): cv.string,
     vol.Required(CONF_ENTITY_IDS): cv.entity_ids,
     vol.Optional(CONF_ROUND_DIGITS, default=2): vol.Coerce(int),
+    vol.Optional(CONF_ICON, default=DEFAULT_ICON): cv.string,
```
Please remove that. Customization of sensors should only be done through `customize:`.

fabaff commented on homeassistant/components/sensor/min_max.py:
```diff
@@ -47,9 +47,11 @@
         vol.All(cv.string, vol.In(SENSOR_TYPES.values())),
-    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_ID): cv.string,
```
This doesn't add real value. The entity IDs are for identifying the entity and should only be set if we are not able to determine it in another way.

micw committed:
Revert "Add option to set entity_id on min_max sensor"

This reverts commit 4685f266477dfaba92cf8a256ecfe62c929c877b.

micw committed:
Revert "Make icon configurable"

This reverts commit fe45aec82d9a07e730bebb4d47dc9618fb695fe6.

micw committed:
Fixes

micw commented:
@fabaff Hope it's fine now?
